gulpfile.js

```diff
@@ -3,111 +3,209 @@ var gulp   = require('gulp');
-  .pipe(gulp.dest(paths.tests.to));
+  .pipe(gulp.dest(paths.test.to));
 });
 
-gulp.task('test', ['compile:tests'], function(done) {
+// Run the unit tests with karma
+gulp.task('test:spec', ['ts:compile:spec'], function(done) {
   karma.start({
-    configFile: __dirname + '/karma.conf.js',
-    singleRun: true
+    configFile: __dirname + '/' + paths.spec.config,
+    singleRun: true,
+    coverageReporter: {
+      type: 'lcovonly',
+      dir: paths.coverage.to
+    }
   }, function() {done()});
 });
 
-gulp.task('e2e', function() {
-  return gulp.src(["./test/e2e/test_utils.js", "./test/e2e/test_cases/**/spec.js"])
-    .pipe($.protractor.protractor({
-      configFile: "./test/e2e/protractor.config.js"
+// Compile the Jinja test templates
+gulp.task('jinja:compile:e2e', function() {
+  // Configure the Nunjucks options
+  $.nunjucksRender.nunjucks.configure(['test']);
+  return gulp.src('test/**/*.tpl.html')
+    .pipe($.nunjucksRender())
+    .pipe($.rename(function(path) {
+      pos = path.basename.search('.tpl');
+      path.basename = path.basename.substring(0, pos);
     }))
-    .on('error', function() {})
+    .pipe(gulp.dest(paths.test.to));
 });
 
-var watchTasks = ['compile:source:concat', 'test', 'e2e'];
+// Serve the static files
+gulp.task('server', $.serve({
+    root: ['.tmp', 'node_modules'],
+    port: 1234
+  })
+);
+
+// Serving files via `gulp serve`
+gulp.task('serve', function(){
+  isWatching = true;
+  gulp.run('server');
+});
+
+// Update webdriver and selenium
+gulp.task('webdriver', webdriver);
+gulp.task('webdriver:update', webdriverUpdate);
+
+// Run the integration tests with protractor
+gulp.task('test:e2e', [
+  'ts:lint:e2e', 'ts:compile:e2e', 'jinja:compile:e2e', 'server'
+], function() {
+  return gulp.src(paths.e2e.from)
+    .pipe(protractor({
+        configFile: paths.e2e.config,
+        keepAlive: false,
+    }))
+    .on('error', function(err) { throw err; });
+});
+
+// Extensive watch on the source and unit test files
+var watchTasks = [
+  'ts:lint:source', 'ts:lint:spec', 'ts:compile:source', 'test:spec'
+];
 gulp.task('watch', watchTasks, function () {
-  gulp.watch([paths.tests.from, paths.source.from, paths.e2e.from], [watchTasks])
+  isWatching = true;
+  gulp.watch([paths.source.from, paths.test.from], [watchTasks])
 });
 
-gulp.task('quick-watch', ['compile:source:concat'], function () {
-  gulp.watch([paths.source.from], [['compile:source:concat']])
+// Quick watch on the source files
+gulp.task('quick-watch', ['ts:compile:source'], function () {
+  isWatching = true;
+  gulp.watch([paths.source.from], [['ts:compile:source']])
 });
 
-gulp.task('coveralls', function() {
-  gulp.src('test/coverage/**/lcov.info')
+// Submit code coverage to Coveralls
+gulp.task('coveralls', ['clean:coverage'], function() {
+  gulp.src(paths.coverage.to + '**/lcov.info')
     .pipe($.coveralls());
 });
 
-gulp.task('default', ['build']);
+// Default task running `gulp`
+gulp.task('default', function(){
```

Comment:
I had some problems, that the gulp task finished nut didnt exit properly. someone proposed this as a workaround but it didnt change anything. So i better change it back :)